service/controller/resource/clusterupgrade/create.go

```diff
@@ -0,0 +1,167 @@
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if !masterUpgraded {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "master node hasn't upgraded yet")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "cancelling resource")
+		return nil
+	}
+
+	err = r.propagateReleaseToMachinePools(ctx, cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensured that machinepools has same release label")
+
+	return nil
+}
+
+func (r *Resource) ensureAzureClusterHasSameRelease(ctx context.Context, cr capiv1alpha3.Cluster) error {
+	if cr.Spec.InfrastructureRef == nil {
+		return microerror.Maskf(notFoundError, "infrastructure reference not yet set")
+	}
+
+	azureCluster := capzv1alpha3.AzureCluster{}
+	err := r.ctrlClient.Get(ctx, client.ObjectKey{Namespace: cr.Namespace, Name: cr.Spec.InfrastructureRef.Name}, &azureCluster)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if cr.Labels[label.ReleaseVersion] == azureCluster.Labels[label.ReleaseVersion] &&
+		cr.Labels[label.AzureOperatorVersion] == azureCluster.Labels[label.AzureOperatorVersion] {
+		// AzureCluster release & operator version already matches. Nothing to do here.
+		return nil
+	}
+
+	azureCluster.Labels[label.AzureOperatorVersion] = cr.Labels[label.AzureOperatorVersion]
+	azureCluster.Labels[label.ReleaseVersion] = cr.Labels[label.ReleaseVersion]
+	err = r.ctrlClient.Update(ctx, &azureCluster)
+	if apierrors.IsConflict(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "conflict trying to save object in k8s API concurrently", "stack", microerror.JSON(microerror.Mask(err)))
+		r.logger.LogCtx(ctx, "level", "debug", "message", "cancelling resource")
+		return nil
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Resource) ensureMasterHasUpgraded(ctx context.Context, cluster capiv1alpha3.Cluster) (bool, error) {
+	tenantClusterK8sClient, err := r.tenantClientFactory.GetClient(ctx, &cluster)
+	if err != nil {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "tenant API not available yet", "stack", microerror.JSON(err))
+		return false, nil
+	}
+
+	nodeList := &corev1.NodeList{}
+	err = tenantClusterK8sClient.List(ctx, nodeList, client.MatchingLabels{"kubernetes.io/role": "master"})
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	for _, n := range nodeList.Items {
+		nodeVer, labelExists := n.Labels[label.AzureOperatorVersion]
+
+		if !labelExists || nodeVer != project.Version() {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func (r *Resource) propagateReleaseToMachinePools(ctx context.Context, cr capiv1alpha3.Cluster) error {
+	lst := expcapiv1alpha3.MachinePoolList{}
+	err := r.ctrlClient.List(ctx, &lst, client.MatchingLabels{capiv1alpha3.ClusterLabelName: key.ClusterName(&cr)})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if !hasMachinePoolsUpgradeCompleted(lst.Items, cr.Labels[label.ReleaseVersion]) {
```

Comment:
Is this the correct place to do this check?

It returns if some `MachinePool` did not finish upgrading, but the upgrade itself is triggered after this in code, when setting the release labels. So if the upgrade check is correct, it will always initially return `false`, as we would never start the upgrade.

I understood we would check if `MachinePool` `n` is upgraded before triggering the upgrade for `MachinePool` `n+1`, but I don't see that here.

Or I'm missing something?

Reply:
Or not...

I mean, this will be new azure-operator that will be running and reconciling CR anyway, so the nodepool handler would just pick up the `AzureMachinePool` CRs anyway and start doing whatever it is doing with them, which includes submitting Azure deployments and doing the entire upgrade.

So in order to enforce the upgrade order, I believe we would need to change `nodepool` resource, so it does not proceed with upgrades until some criteria is met.

So what you implemented here, setting release and azure-operator labels, that would trigger the setting of `Upgrading` to `True` (what I have in another branch), and then in `nodepool` handler we need to check if `Upgrading` condition is set, and proceed with upgrade only it is set to `True`, otherwise wait in the same state.

Reply:
I don't fully follow.

My strategy / thinking here is following:

- First ensure that CP gets upgraded. This is currently ensured by checking master nodes' `azure-operator` version label from tenant api directly. Slightly ugly, but works for now until we get specific types for CP.
- Then proceed to node pools as follows:
  - To proceed, all `MachinePool` CRs must have `Upgrading` condition false for to-be-upgraded release version. We also need to check here if labels were already updated (i.e. `machine-pool-controller` didn't pick up the CR update yet).
  - When no node pool is upgrading, pick one and update labels.
    - Now `machine-pool-controller` gets event (eventually), sets the `Upgrading` condition and reflects the change to `AzureMachinePool`.
      - `azure-machine-pool` controller is straightforward and works by detecting drift in desired infrastructure vs. present infrastructure and therefore it proceeds upgrading the VMSS.

My reasoning for this is that upgrade strategy should be controlled on abstract high level since it's nothing provider specific. I'd also strongly discourage towards dealing with sibling CRs because for given controller they are fully different things working in different context and coordination between them is responsibility of component that manages them, by that very definition. There are few exceptions to this, such as IPAM where installation-specific uniqueness must be guaranteed, but even that one should probably be implemented by a separate operator.

Does that make any sense?

Reply:
Maybe we can talk tomorrow about the whole flow, it's not easy to understand. At the end I'd like to have a comment in the code explaining the whole thing.